mcorpc/authz_rego.go

```diff
@@ -0,0 +1,131 @@
+package mcorpc
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+
+	"github.com/choria-io/go-choria/choria"
+	"github.com/choria-io/go-config"
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/rego"
+	"github.com/open-policy-agent/opa/topdown"
+	"github.com/sirupsen/logrus"
+)
+
+type regoPolicy struct {
+	cfg   *config.Config
+	req   *Request
+	agent *Agent
+	log   *logrus.Entry
+}
+
+func regoPolicyAuthorize(req *Request, agent *Agent, log *logrus.Entry) (bool, error) {
+	logger := log.WithFields(logrus.Fields{
+		"authorizer": "regoPolicy",
+		"agent":      agent.Name(),
+		"request":    req.RequestID,
+	})
+
+	authz := &regoPolicy{
+		cfg:   agent.Config,
+		req:   req,
+		agent: agent,
+		log:   logger,
+	}
+
+	return authz.authorize()
+}
+
+func (r *regoPolicy) authorize() (bool, error) {
+	policyFile, err := r.lookupPolicyFile()
+	if err != nil {
+		return false, err
+	}
+
+	pol, err := ioutil.ReadFile(policyFile)
+	if err != nil {
+		return false, err
+	}
+
+	module := string(pol)
+
+	buf := topdown.NewBufferTracer()
+
+	options := []func(*rego.Rego){
+		rego.Query("data.choria.mcorpc.authpolicy.allow"),
+		rego.Module("//choria/mcorpc/policies/"+r.agent.meta.Name, module),
+		rego.Input(r.regoInputs()),
+	}
+
+	if r.log.Logger.GetLevel() == logrus.TraceLevel {
+		options = append(options, rego.Tracer(buf))
+	}
+
+	query := rego.New(
+		options...,
+	)
+
+	rs, err := query.Eval(context.Background())
+
+	if r.log.Logger.GetLevel() == logrus.TraceLevel {
+		topdown.PrettyTrace(r.log.Writer(), *buf)
+	}
+
+	switch err := err.(type) {
+	case nil:
+		break
+	case ast.Errors:
+		for _, e := range err {
+			r.log.Info("code: ", e.Code)
+			r.log.Info("row: ", e.Location.Row)
+			r.log.Info("filename: ", policyFile)
+		}
+	default:
+		return false, err
+	}
+
+	return rs[0].Expressions[0].Value.(bool), nil
+}
+
+func (r *regoPolicy) lookupPolicyFile() (string, error) {
+	dir := filepath.Join(filepath.Dir(r.cfg.ConfigFile), "policies", "rego")
+
+	regoPolicy := filepath.Join(dir, r.agent.Name()+".rego")
+
+	r.log.Debugf("Looking up rego policy in %s", regoPolicy)
+	if choria.FileExist(regoPolicy) {
+		return regoPolicy, nil
+	}
+
+	defaultPolicy := filepath.Join(dir, "default.rego")
+	if choria.FileExist(defaultPolicy) {
+		return defaultPolicy, nil
+	}
+	return "", fmt.Errorf("no policy %s found for %s in %s", defaultPolicy, r.agent.Name(), dir)
+
+}
+
+func (r *regoPolicy) regoInputs() map[string]interface{} {
+	facts := map[string]interface{}{}
+
+	err := json.Unmarshal(r.agent.ServerInfoSource.Facts(), &facts)
+	if err != nil {
+		r.log.Errorf("unable to unmarshal facts?")
```

Comment:
`"could not marshal facts for rego policy: %v", err`

I think we should fail when facts cant marshal, dont want to in that situation accidentally allow something that a policy would deny if all the facts were present.

that implies a few things though: 

 - we should check if the facts file doesnt exist and skip it without failing
 - we should check if its 0 bytes and not fail
 - then if it doesnt decode we fail

Also though facts can be yaml or JSON thus needs converting, so use https://github.com/choria-io/go-protocol/blob/3b4d7c391d1a22a2c6c20b10f751c98054831a23/filter/facts/facts.go#L55 to do the reading it will do all the right things with conversion etc - but do the exist and size check first

Reply:
Given that this is set by facts.JSON in go-protocol/filter/facts, I don't thin I need to do that?  It should at least by an empty document.

Reply:
problem is it squashes errors, so a corrupt facts file can lead to empty facts.

Lets at least check for FactsPresent && FactNotZeroSize && len(JSON) > 0 - so we know a none zero length facts file always yields facts and if not we fail requests to it?

Reply:
Still kind of unsure what the threat of empty facts would be.

Ie, if there's no facts set, and you're depending on them to be set in your policy, it would reject the authorization?

If the contents were corrupted on disk, the facts.JSON() function would just return "{}"..

Reply:
ok, lets leave it for now then